core/src/docs/rfcs/5444_operator_from_uri.md

```diff
@@ -0,0 +1,144 @@
+- Proposal Name: `operator_from_uri`
+- Start Date: 2024-12-23
+- RFC PR: [apache/opendal#5444](https://github.com/apache/opendal/pull/5444)
+- Tracking Issue: [apache/opendal#5445](https://github.com/apache/opendal/issues/5445)
+
+# Summary
+
+This RFC proposes adding URI-based configuration support to OpenDAL, allowing users to create operators directly from URIs. The proposal introduces a new `from_uri` API in both the `Operator` and `Configurator` traits, along with an `OperatorRegistry` to manage operator factories.
+
+# Motivation
+
+Currently, creating an operator in OpenDAL requires explicit configuration through builder patterns. While this approach provides type safety and clear documentation, it can be verbose and inflexible for simple use cases. Many storage systems are naturally identified by URIs (e.g., `s3://bucket/path`, `fs:///path/to/dir`).
+
+Adding URI-based configuration would:
+
+- Simplify operator creation for common use cases
+- Enable configuration via connection strings (common in many applications)
+- Make OpenDAL more approachable for new users
+- Allow dynamic operator creation based on runtime configuration
+
+# Guide-level explanation
+
+The new API allows creating operators directly from URIs:
+
+```rust
+// Create an operator using URI
+let op = Operator::from_uri("s3://my-bucket/path", vec![
+    ("access_key_id".to_string(), "xxx".to_string()),
```

Comment:
Why do we need this additional parameters for key-value secrets?

I think we should only use the uri. If secrets need to be configured, they can be set in the uri itself.

If you "hardcode" this pair of secrets, the runtime configuration changes would be more difficult, as the vector of pairs would be tied to an specific storage backend.

Reply:
> Why do we need this additional parameters for key-value secrets?

Hi, there are many cases where users need additional options. For example, applications can have `SERVICE_ACCESS_KEY_ID` and `SERVICE_SECRET_ACCESS_KEY`. Users may input strings such as `s3://bucket_a` or `s3://bucket_b/path/to/dir`. Applications no longer need to find ways to handle these URIs.

The additional options also make it easier for developers to include extra configurations beyond the URI itself, which might be difficult to convey or require URL encoding/decoding.

> I think we should only use the uri. If secrets need to be configured, they can be set in the uri itself.

Yes, user can pass ak/sk in that way.

> If you "hardcode" this pair of secrets, the runtime configuration changes would be more difficult, as the vector of pairs would be tied to an specific storage backend.

It is a fact that different services have different configurations. Addressing this is not within the scope of `Operator::from_uri`. Users need to provide the configuration in one of several ways: through the URI, options, environment variables, etc.

In the best cases, OpenDAL can correctly load secrets from the environment. This allows users to simply use `s3://bucket` or `azblob://container`, depending on their own environment setup.

Reply:
> In the best cases, OpenDAL can correctly load secrets from the environment. This allows users to simply use s3://bucket or azblob://container, depending on their own environment setup. But it's not the goal of from_uri to address.

Oh, okay, didn't know that! I think as long as users can switch backend with different env vars, it will be fine.

thank you very much for your explanation, this is very clear to me now